Isosurfacing_3/include/CGAL/Isosurfacing_3/edge_intersection_oracles_3.h

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2022-2024 INRIA Sophia-Antipolis (France), GeometryFactory (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Julian Stahl
+//                 Mael Rouxel-Labbé
+
+#ifndef CGAL_ISOSURFACING_3_EDGE_INTERSECTION_ORACLES_3_H
+#define CGAL_ISOSURFACING_3_EDGE_INTERSECTION_ORACLES_3_H
+
+#include <CGAL/license/Isosurfacing_3.h>
+
+#include <CGAL/assertions.h>
+
+namespace CGAL {
+namespace Isosurfacing {
+
+/**
+ * \ingroup IS_Domain_helpers_grp
+ *
+ * \cgalModels{IsosurfacingEdgeIntersectionOracle_3}
+ *
+ * \brief The class `Dichotomy_edge_intersection` uses a dichotomy to find the intersection point
+ * between an edge and the isosurface.
+ *
+ * This class is for example suitable to be used as the `EdgeIntersectionOracle` template parameter of isosurfacing
+ * domain classes when values are computed using an implicit function.
+ * It is however not optimal when the values are interpolated from discrete values
+ * since the intersection can be computed analytically in this case.
+ *
+ * \sa `CGAL::Isosurfacing::Linear_interpolation_edge_intersection`
+ * \sa `CGAL::Isosurfacing::Marching_cubes_domain_3`
+ * \sa `CGAL::Isosurfacing::Dual_contouring_domain_3`
+ * \sa `CGAL::Isosurfacing::Value_field_3`
+ */
+struct Dichotomy_edge_intersection
+{
+  unsigned int m_max_iterations = 10;
+  double m_relative_eps = 1e-7;
+
+  /*!
+   * \brief computes the intersection point between an edge and the isosurface.
+   *
+   * \tparam Domain must be a model of `IsosurfacingDomain_3`
+   *
+   * \param p_0 the geometric position of the first vertex of the edge
+   * \param p_1 the geometric position of the second vertex of the edge
+   * \param val_0 the value at the first vertex of the edge
+   * \param val_1 the value at the second vertex of the edge
+   * \param domain the isosurfacing domain
+   * \param isovalue the isovalue defining the isosurfacing with which we seek an intersection
+   * \param p the intersection point, if it exists
+   *
+   * \return `true` if the intersection point exists, `false` otherwise
```

Comment:
return `std::optional<Point_3>`?

Reply:
Internally we use the value `p` to store intermediate computations of the dichotomy. It could be done with a local point (and probably without overhead if things are moved), but I think it would make code less readable.